File: leet-code/src/problems/p0020.rs
//[COMPLETED]

struct Solution;

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let (mut paren, mut curly, mut square) = (0i32, 0i32, 0i32);
        if s.len() == 0 { return true; }
        let mut stack: Vec<char> = Vec::new();
        for c in s.chars() {
//            println!("{}", c);
            match c {
                // Open
                '(' => { paren += 1;  stack.push(c); },
                '[' => { square += 1; stack.push(c); },
                '{' => { curly += 1;  stack.push(c); },
                // Close
                ')' => {
                    match stack.pop() {
                        Some(p) => { if p != '(' { return false; } },
                        None => { return false}
                    }
                    paren -= 1;
                },
                ']' => {
                    match stack.pop() {
                        Some(p) => { if p != '[' { return false; } },
                        None => { return false}
                    }
                    square -= 1;
                },
                '}' => {
                    match stack.pop() {
                        Some(p) => { if p != '{' { return false; } },
                        None => { return false}
                    }
                    curly -= 1;
                },
                _ => return false,  // Unrecognized character
            }
        }
        paren == 0 && curly == 0 && square == 0
    }
}

fn main() {
    println!("{}", Solution::is_valid("[()({}{}[]())]".to_string()));
}


// Bad try, need a stack
//pub fn is_valid(s: String) -> bool {
//    let (mut paren, mut curly, mut square) = (0i32, 0i32, 0i32);
//    if s.len() == 0 { return true; }
//    let mut last_open: char = ' ';  // Fails if first character is closing
//    for c in s.chars() {
//        println!("{}", c);
//        match c {
//            // Open
//            '(' => { paren += 1;  last_open = c},
//            '[' => { square += 1; last_open = c},
//            '{' => { curly += 1;  last_open = c},
//            // Close
//            ')' => {
//                if last_open != '(' { return false }
//                paren -= 1;
//            },
//            ']' => {
//                if last_open != '[' { return false }
//                square -= 1;
//            },
//            '}' => {
//                if last_open != '{' { return false }
//                curly -= 1;
//            },
//            _ => return false,
//        }
//    }
//    paren == 0 && curly == 0 && square == 0
//}

File: leet-code/problems/p0020.java
//[COMPLETED]

package p020;

import java.util.List;
import java.util.Stack;

public class P020 {

    public static void main(String[] args) {
        Solution s = new Solution();
        System.out.println(s.isValid(""));
        System.out.println(s.isValid("()"));
        System.out.println(s.isValid("()[]{}") + "\t()[]{}");
        System.out.println(s.isValid("(]"));
        System.out.println(s.isValid("([)]"));
    }
    
    
private static class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        
        for (char c : s.toCharArray() ) {
            if (c == '(' || c == '{' || c == '[')
                stack.add(0, c);
            if (c == ')') {
                if (stack.isEmpty()) return false;
                if (stack.remove(0) != '(') return false;
            }
            if (c == ']') {
                if (stack.isEmpty()) return false;
                if (stack.remove(0) != '[') return false;
            }
            if (c == '}') {
                if (stack.isEmpty()) return false;
                if (stack.remove(0) != '{') return false;
            }
        }
        
        return stack.isEmpty();
    }
}

}
